### PR TITLE
A new build process, based on GitHub Actions

### PR DESCRIPTION
**Pull request by apparentlymart**
This introduces a new process, based on GitHub Actions, for producing the various release artifacts for Terraform CLI.

The _ultimate_ goal here is for this workflow to build all of the artifacts needed to drive the Terraform CLI release process, so that the actual release process (maintained elsewhere) could consume these artifacts directly and no longer need to access the actual repository content, which in turn means that we can verify all of the artifacts before optionally launching the separate release process. The intended workflow then would be to first have this workflow build on the release branch, then once it's succeeded create a tag from the same commit to produce a releasable build. The branch build will have a synthetic version number including the commit ID, while the _tag_ build would reflect the version number indicated in the tag.

However, **for now my goals here are more modest**: just to get a new process in place which will give us earlier warning if anything merged to a release branch or included in a tagged version has any build problems. No automated process is actually consuming the results of this for now, and I expect we'll iterate more on this workflow once we're ready to integrate this with a new release process. Which is to say: _there are some known and documented limitations here_, which we'll hopefully address in later commits, but which are not in scope for this PR.

The effect of merging this will be that any future merges to a release branch or to the `main` branch will trigger a full build for all of our target platforms, including the construction of the various "wrapper package" types our official process produces (Deb packages, RPM packages, Docker Images). It will then run the end-to-end tests against the `linux_amd64`, `darwin_amd64`, and `windows_amd64` packages, which should increase our confidence that we built generally-working packages across those three platforms.

Given that, I'd like to merge this to get that relatively-minor benefit for now and also to get some experience with the behavior of this workflow in advance of us integrating it with the rest of the release process. In particular, we'll be able to see how it behaves in our _real_ workflow with release branches and tagging, whereas so far I've only been able to test it on this development branch.

---

Because for the moment we will not actually be releasing the packages built by this process (we'll continue to release the artifacts produced by the non-public release workflow as part of releasing), this doesn't technically assure us that the _released_ packages would also pass the e2etests once built, but in practice the build process here is similar enough to the one the real release process uses that it should still serve as a good early warning signal.

The later work to integrate this with a new release process that actually _consumes_ all of these artifacts will be in closer collaboration with the release engineering team and I expect it will result in at least some small changes to the details of how this works, so we can make sure this is producing artifacts in a suitable shape for the standard release process. However, for my part I did derive this from the build process templates provided from the release engineering team, and referred to the corresponding internal documentation, and so I believe it should already be broadly compatible with those assumptions and only require minor tweaks later.

---

Although this PR doesn't modify any "real" Terraform code, it does include some adjustments to the `internal/command/e2etest` package to make it ready to support the separate build run approach used by this workflow. This separation between building the test harness and running it was an original intention of the e2etest design, to allow us to be sure that we're really testing the actual release executable and not a fresh build from source, but we previously didn't end up actually using it in that mode and so that capability has "bitrotted" a bit over the last few years.

In particular, we introduced a few new tests that expect to be able to build new executables from source _during the test run_, which is counter to the original idea of a test suite we can run against pre-built executables. As a compromise for now I've just made the test harness skip over those tests when running in this mode, since we'll still get those run _on Linux_ during the normal pull request check process. In a later commit I'd like to adjust this slightly so that the `make_archive.sh` script _also_ includes the needed supporting executables and thus we'll be able to make the separate test harness run the full test suite as originally intended.

---

You can preview the effect of this workflow by viewing [the Actions runs generated from this workflow on this development branch](https://github.com/hashicorp/terraform/actions/workflows/build.yml?query=branch%3Abuild-workflow-dev%2Fmatkins-initial), though I believe you'll need to be a maintainer in this repository in order to be able to actually download the built artifacts.

Note that there's a quirk in the GitHub Actions UI whereby any artifact you download is always wrapped in a `.zip` archive container, _even if the artifact itself is a `.zip` file_. That's kinda annoying for the various `.zip` packages we produce here, but is an expected part of how GitHub Actions works and doesn't really hurt for our final intended use because the GitHub Actions steps for consuming these artifacts downstream automatically unpacks the outer wrapping archive to reveal the _real_ `.zip` file inside. This extra level of wrapping is actually important because GitHub Actions's _own_ packaging cannot preserve file permissions, and so the extra inner `.zip` file we produced ourselves can preserve the executable permission on the `terraform` executable.

### CODE REVIEW

File: internal/command/e2etest/provider_dev_test.go
Comment (alisdair) [inline internal/command/e2etest/provider_dev_test.go:30]: Unless I'm misunderstanding something, these skipped e2e tests don't seem to run in CI at all any more. They seem to be skipped both in the new build-then-run actions and in the per-commit CircleCI runs.

If that's the case, is it feasible to fix that before merge?

**Review by alisdair — APPROVED** [review]
Approved with one concern about the newly skipped e2e tests

File: internal/command/e2etest/provider_dev_test.go
Comment (apparentlymart) [reply internal/command/e2etest/provider_dev_test.go:30]: Ahh yes, you are right that the "normal" path which builds a custom binary sneakily updates this variable before it returns:

https://github.com/hashicorp/terraform/blob/535da4ebc754a7b246c3314aeeeceab43b0ac854/internal/command/e2etest/main_test.go#L21-L39

I'll fix this up by adding a separate variable to track whether we're using an externally-provided executable or a locally-built one, and then change these conditions to all depend on that instead, as a more robust temporary solution to this quirk. Thanks!

File: .github/workflows/build.yml
Comment (morganchristiansson) [inline .github/workflows/build.yml:271]: There's also qemu. It is significantly slower as it's cpu instruction emulation. But for low-cpu tasks it works great.

Just use this github action and then you can exec both amd64 and arm64 binaries https://github.com/docker/setup-qemu-action

Comment (morganchristiansson) [reply .github/workflows/build.yml:271]: I can have a go and updating package-docker to build arm64 docker image and and e2etest-* to test arm64 binary using amd64 test harness if that sounds ok?

Comment (apparentlymart) [reply .github/workflows/build.yml:271]: Hi @morganchristiansson,

Thanks for the tip!

I believe the packaging helper action we're using here already has some support for packaging across architectures -- I think I saw it work when I was first experimenting. However, I intentionally reduced it to just what our current release pipeline is capable of so that we wouldn't get potentially blocked by irrelevant failures in the meantime, given that we're currently using this only to get early feedback on build/e2etest problems.

I'd like to wait to add new architectures here until our release engineering team is ready to work with us on feeding this into the rest of the new pipeline, so that we can make sure whatever we do here is consistent with how other projects using the same pipeline will do it.

Comment (morganchristiansson) [reply .github/workflows/build.yml:271]: Btw I had a look at https://github.com/hashicorp/actions-docker-build it already uses `docker build buildx --platform $PLATFORM` but $PLATFORM is only used internally not settable from action parameters. https://github.com/hashicorp/actions-docker-build/blob/v1/action.yml#L222-L224`

If arm64 binary is being built here? It will be easy to add the binary to arm64 docker image like this and will run well...

**Pull request merged**
